practice_2/NewsApp/Presenter/NewsPresenter.swift

```diff
@@ -0,0 +1,135 @@
+import Foundation
+
+let newsCacheExpired = 1  // 15 minutes
+let newsMaxNumberOfConnectionAttempts = 5
+
+protocol NewsViewProtocol: AnyObject {
+    func failure(error: Error)
+}
+
+protocol NewsPresenterProtocol: AnyObject {
+    var topArticles: [TopArticle] { get set }
+    init(view: NewsViewProtocol, networkService: NewsNetworkServiceProtocol)
+    func fetchArticles(by keyword: String)
+    func sortAlphabetically(inOrder: String)
+}
+
+class NewsPresenter: NewsPresenterProtocol {
+    var topArticles = [TopArticle]()
+    weak var view: NewsViewProtocol?
+    let networkService: NewsNetworkServiceProtocol
+    let dbService = DefaultNewsDBService(context: NewsDatabaseStack.persistentContainer.viewContext)
+
+    required init(view: NewsViewProtocol, networkService: NewsNetworkServiceProtocol) {
+        self.view = view
+        self.networkService = networkService
+        getTopNews()
+    }
+
+    private func getTopNews() {
+        let now = NSDate()
+        let lastApiDate = dbService.getDateFromDB()
+        let timeSinceLastApiUpdate = Int(now.timeIntervalSince(lastApiDate as Date))  // in seconds
+
+        if timeSinceLastApiUpdate > newsCacheExpired {
+            dbService.deleteAllArticlesFromDatabase()
+            getArticlesFromApi()
+        } else {
+            print("Get data from DB")
+            dbService.getArticlesFromDatabase(onComplete: { articlesDB in
+                if articlesDB.count == 0 {
```

Comment:
do not compare to 0, use .isEmpty or something like that

Reply:
Thank you.